crates/turbopack-build/src/minify.rs

```diff
@@ -0,0 +1,145 @@
+
+#[turbo_tasks::function]
+pub async fn minify(path: Vc<FileSystemPath>, code: Vc<Code>) -> Result<Vc<Code>> {
+    let original_map = *code.generate_source_map().await?;
+    let minified_code = perform_minify(path, code);
+
+    let merged = match (original_map, *minified_code.generate_source_map().await?) {
+        (Some(original_map), Some(minify_map)) => Some(Vc::upcast(
+            TracedSourceMap(original_map.trace(minify_map)).cell(),
+        )),
+        _ => None,
+    };
+
+    let mut builder = CodeBuilder::default();
+    builder.push_source(minified_code.await?.source_code(), merged);
+    let path = &*path.await?;
+    let filename = path.file_name();
+    write!(builder, "\n\n//# sourceMappingURL={}.map", filename)?;
+    return Ok(builder.build().cell());
+}
+
+#[turbo_tasks::function]
+async fn perform_minify(path: Vc<FileSystemPath>, code_vc: Vc<Code>) -> Result<Vc<Code>> {
+    let code = &*code_vc.await?;
+    let cm = Arc::new(SwcSourceMap::new(FilePathMapping::empty()));
+    let compiler = Arc::new(Compiler::new(cm.clone()));
+    let fm = compiler.cm.new_source_file(
+        FileName::Custom((*path.await?.path).to_string()),
+        code.source_code().to_str()?.to_string(),
+    );
+
+    let lexer = ecma::parser::lexer::Lexer::new(
+        ecma::parser::Syntax::default(),
+        ecma::ast::EsVersion::latest(),
+        ecma::parser::StringInput::from(&*fm),
+        None,
+    );
+    let mut parser = ecma::parser::Parser::new_from(lexer);
+    let program = try_with_handler(cm.clone(), Default::default(), |handler| {
+        GLOBALS.set(&Default::default(), || {
+            let program = parser.parse_program().unwrap();
+            let unresolved_mark = Mark::new();
+            let top_level_mark = Mark::new();
+
+            Ok(compiler.run_transform(handler, false, || {
+                swc_core::ecma::minifier::optimize(
+                    program,
+                    cm.clone(),
+                    None,
+                    None,
+                    &ecma::minifier::option::MinifyOptions {
+                        compress: Some(Default::default()),
+                        ..Default::default()
+                    },
+                    &ecma::minifier::option::ExtraOptions {
+                        top_level_mark,
+                        unresolved_mark,
+                    },
+                )
+            }))
+        })
+    })?;
+
+    let (src, src_map_buf) = emit(cm.clone(), program)?;
+
+    let mut builder = CodeBuilder::default();
+    builder.push_source(
+        &src.into(),
+        Some(*Box::new(Vc::upcast(
+            ParseResultSourceMap::new(cm, src_map_buf).cell(),
+        ))),
+    );
+
+    Ok(builder.build().cell())
+}
+
+// From https://github.com/swc-project/swc/blob/11efd4e7c5e8081f8af141099d3459c3534c1e1d/crates/swc/src/lib.rs#L523-L560
+fn emit(cm: Arc<SwcSourceMap>, program: Program) -> Result<(String, Vec<(BytePos, LineCol)>)> {
```

Comment:
Nit: This is usually called `print`, and I was initially confused why we were emitting collectibles into the task graph.